lib/dialects/abstract/query-generator.js

```diff
@@ -185,9 +185,25 @@ class QueryGenerator {
         // If no conflict target columns were specified, use the primary key names from options.upsertKeys
         const conflictKeys = options.upsertKeys.map(attr => this.quoteIdentifier(attr));
         const updateKeys = options.updateOnDuplicate.map(attr => `${this.quoteIdentifier(attr)}=EXCLUDED.${this.quoteIdentifier(attr)}`);
-        onDuplicateKeyUpdate = ` ON CONFLICT (${conflictKeys.join(',')}) DO UPDATE SET ${updateKeys.join(',')}`;
+        onDuplicateKeyUpdate = ` ON CONFLICT (${conflictKeys.join(',')})`;
+        // if update keys are provided, then apply them here.  if there are no updateKeys provided, then do not try to
+        // do an update.  Instead, fall back to DO NOTHING.
+        onDuplicateKeyUpdate += updateKeys.length ? ` DO UPDATE SET ${updateKeys.join(',')}` : ' DO NOTHING';
       } else {
         const valueKeys = options.updateOnDuplicate.map(attr => `${this.quoteIdentifier(attr)}=VALUES(${this.quoteIdentifier(attr)})`);
```

Comment:
I have a little bit of a hard time understanding what updateOnDuplicate is and what it is used for.

Reply:
It contains the list of fields (just the field names, no values) that will be updated when an insert fails with a duplicate error.

So if a developer called something like this:

```
const [ instance ] = await MyModel.upsert({ id: 1, email: 'xxx@yyy.zzz', phone: '123456' }, { fields: [ 'phone' ], returning: true });
```

The `upsert` query-interface function get's called with something like this:

```
await upsert('MyModel', { id: 1, email: 'xxx@yyy.zzz', phone: '123456' }, { phone: '123456' }, { where: { id: 1 } }, options);
```

It gets set by the abstract query interface here: https://github.com/sequelize/sequelize/blob/199b632b021830f9d09210fd7430045710638631/lib/dialects/abstract/query-interface.js#L775

The `upsert` function overridden by the mysql interface here: https://github.com/sequelize/sequelize/blob/7006eb4ebbfd38cc055873f56c6593f633f28bad/lib/dialects/mysql/query-interface.js#L48

So by the time it gets to this code, the in the `query-generator`, the `updateOnDuplicate` option is set to `[ 'phone' ]`.

I'm adding a comment in the code as I respond to the rest of your code review comments.

Reply:
The terminology (and logic for that matter) for upserts (being `Model.bulkCreate` and `Model.upsert`) in general isn't very consistent I've found.

`updateOnDuplicate` for `Model.bulkCreate` serves the same purpose as `firelds` for `Model.upsert`.

There's additionally an option for `Model.bulkCreate`, `ignoreDuplicates` which translates to `ON CONFLICT DO NOTHING`, similar to what this pr will change `upsert` to do implicitly if none of the specified fields are provided.